hibike/hibike_message.py

```diff
@@ -112,31 +131,195 @@ def checksum(data):
+
+  return message
+
+def make_device_read(device_id, params):
+  """ Makes and returns DeviceRead message.
+
+      looks up config data about the specified 
+      device_id to properly construct the message.
+      
+      device_id - a device type id (not uid).
+      params    - an iterable of param names
+  """
+  paramNums = [paramMap[device_id][name][0] for name in params]
+  entries = [1 << num for num in paramNums]
+  tot = 0
+  for i in range(len(entries)):
+    tot = tot ^ entries[i]
+  temp_payload = struct.pack('<H', tot)
+  payload = bytearray(temp_payload)
+  message = HibikeMessage(messageTypes["DeviceRead"], payload)
+  return message
+
+def decode_params(device_id, params):
+#     Decodes an inputted set of parameters that is in binary form   
+#     Returns a list of names symbolizing the parameters encoded 
+#
+#     device_id - a device type id (not uid)
+#     params    - the set of parameters in binary form
+  converted_params = []
+  for param_count in range(16):
+     if (1 & (params >> param_count) == 1):
+        converted_params.append(param_count)
+  named_params = []
+  for param in converted_params:
+     if param >= len(devices[device_id]["params"]):
+        break
+     named_params.append(devices[device_id]["params"][param]["name"])
+  return named_params
+  
+def make_device_write(device_id, params_and_values):
+  """ Makes and returns DeviceWrite message.
+      If all the params cannot fit, it will fill as many as it can.
+
+      looks up config data about the specified 
+      device_id to properly construct the message.
+      
+      device_id         - a device type id (not uid).
+      params_and_values - an iterable of param (name, value) tuples
+  """
+  params_and_values = sorted(params_and_values, key=lambda x: paramMap[device_id][x[0]][0])
+  params = [param[0] for param in params_and_values]
+  paramNums = [paramMap[device_id][name][0] for name in params]
+  paramT = [paramMap[device_id][name][1] for name in params]
+  values = [param[1] for param in params_and_values]
+  entries = [1 << num for num in paramNums]
+
+  tot = 0
+  for i in range(len(entries)):
+    tot = tot ^ entries[i]
+  typeString = '<H'
+  for t in paramT:
+    typeString += paramTypes[t]
+  temp_payload = struct.pack(typeString, tot, *values)
   payload = bytearray(temp_payload)
-  message = HibikeMessage(messageTypes["DeviceUpdate"], payload)
+  message = HibikeMessage(messageTypes["DeviceWrite"], payload)
   return message
 
-def make_device_status(param):
-  """ Makes and returns DeviceStatus message."""
-  temp_payload = struct.pack('<B', param)
+def old_decode_device_write(device_id, message):
+  messageID = message.getmessageID()
+  payload = message.getPayload()
+  messageT = "DeviceWrite"
+  paramB = bin(payload[0])
+  paramNames = decode_params(device_id, paramB)
+  paramT = [paramMap[device_id][name][1] for name in paramNames]
+  typeString = '<H'
+  for t in paramT:
+    typeString += paramTypes[t]
```

Comment:
Construction of a type string from param types should be abstracted into a function.